### PR TITLE
(broken) inital setup of jest Mock

### PR DESCRIPTION
**Pull request by ethanburrell**
Hello,

I am trying to set up a ‘Mock’ class to import via ‘bun:test’, and was wondering on how to setup the scaffolding.

I think I’ve set up the correct pieces, but can’t get the Mock class to be imported by a test.

I set this up by modifying ‘jest.classes.ts’, set up the struct in ‘jest.zig’, and then run ‘make codegen’ and ‘make build’.

Would appreciate any direction.

### CODE REVIEW

**Comment by Jarred-Sumner** [comment]
This is very close but the remaining thing is to wire up Mock to be returned by Bun.jest(). The "bun:test" import doesn't actually exist right now. The JS printer has a transform which converts imports from "bun:test" to Bun.jest(import.meta.path)

So the Bun.jest function needs to return the mock function

File: src/bun.js/test/jest.zig [hunk 1]
```diff
@@ -1692,6 +1742,16 @@ pub const DescribeScope = struct {
...
     }
 
+    pub fn createFn(
+        _: *DescribeScope,
+        ctx: js.JSContextRef,
+        _: js.JSValueRef,
+        _: js.JSStringRef,
+        _: js.ExceptionRef,
+    ) js.JSObjectRef {
+        return js.JSObjectMake(ctx, Mock.fn2.get().*, null);
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: This isn't quite right

Instead

```zig
return Mock.getConstructor(ctx).asObjectRef();
```

File: src/bun.js/test/jest.zig [hunk 2]
```diff
@@ -1384,6 +1384,55 @@ pub const TestScope = struct {
...
+            .count = 0,
+        };
+        return mock;
+    }
+
+    pub fn fn1(_: *Mock, _: *JSC.JSGlobalObject, _: *JSC.CallFrame) callconv(.C) JSValue {
+        return JSValue.zero;
+    }
+
+    pub const fn2 = NewClass(Mock, .{ .name = "fn" }, .{
```
Comment (Jarred-Sumner) [inline src/bun.js/test/jest.zig]: We don't want to use `NewClass` here

`NewClass` is the previous iteration of Bun's JavaScriptCore bindings. That one uses the (slower) JavaScriptCore C API. We haven't migrated all the existing code paths to it, which is why you see it in use still.

`JSC.Codegen.JSMock` is the newer bindings and its the one to use here

File: src/bun.js/test/jest.zig (hunk 1 above)
Comment (ethanburrell) [reply src/bun.js/test/jest.zig]: Thanks for the pointer @Jarred-Sumner.

I have one issue when I use `Mock.getConstructor`. During the run of `make dev` I get the error during the `bun-link-lld-debug` phase:

```
ld.lld: error: undefined symbol: Mock__getConstructor
>>> referenced by generated_classes.zig:2150 (/build/bun/bun/src/bun.js/bindings/generated_classes.zig:2150)
>>>               /build/bun/bun/packages/debug-bun-linux-x64//bun-debug.o:(.javascript_core.bun.js.bindings.generated_classes.JSMock.getConstructor)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:1418: bun-link-lld-debug] Error 1
```

I tried adding / running it with `-v` and was not able to get a better error message. Is there any piece that I'm missing why the link to the constructor would not exist?

Comment (Jarred-Sumner) [reply src/bun.js/test/jest.zig]: you'll need to run `make bindings -j12` to rebuild C++

**Comment by Jarred-Sumner** [comment]
Note that returning JSValue.zero is a crash. It's a value which is supposed to be unreachable by JS.

File: src/bun.js/test/jest.zig (hunk 1 above)
Comment (ethanburrell) [reply src/bun.js/test/jest.zig]: @Jarred-Sumner I'm seeing another problem sometimes after I run `make codegen`, `make bindings`, and then this error happens after `make dev`.

```
ld.lld: error: undefined symbol: ubrk_following_70
>>> referenced by StringView.cpp
>>>               StringView.cpp.o:(WTF::StringView::GraphemeClusters::Iterator::Iterator(WTF::StringView, unsigned int)) in archive /build/bun-webkit/lib/libWTF.a
>>> referenced by StringView.cpp
>>>               StringView.cpp.o:(WTF::StringView::GraphemeClusters::Iterator::operator++()) in archive /build/bun-webkit/lib/libWTF.a
>>> referenced by UnifiedSource-f2e18ffc-17.cpp
>>>               UnifiedSource-f2e18ffc-17.cpp.o:(JSC::IntlSegments::containing(JSC::JSGlobalObject*, JSC::JSValue)) in archive /build/bun-webkit/lib/libJavaScriptCore.a

...

ld.lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:1422: bun-link-lld-debug] Error 1
```

There are many other issues identifying this similar symbols. I've tried cleaning, and re making the headers and bindings. Am I missing anything? (Working in a devcontainer with dependencies and the new zig compiler from `main`).

Any recommended steps to fix this?

**Comment by Electroid** [comment]
Thanks for taking a stab at this, we've since implemented mocks.

**Pull request closed without merging**